gendox-core-api/src/main/java/dev/ctrlspace/gendox/spring/batch/jobs/training/steps/DocumentInstanceSectionProcessor.java

```diff
@@ -68,7 +83,28 @@ public SectionEmbeddingDTO process(DocumentInstanceSection item) throws Exceptio
 
             logger.trace("Section value with template for embedding: {}", sectionValue);
 
-            embeddingResponse = embeddingService.getEmbeddingForMessage(projectAgent, sectionValue, projectAgent.getSemanticSearchModel());
+
+            if (Boolean.TRUE.equals(reuseEmbeddings)) {
+                EmbeddingGroup embeddingGroup = embeddingService.findBySectionOrMessage(item.getId(), null, projectAgent.getSemanticSearchModel().getId());
+
+                String sectionSha256Hash = cryptographyUtils.calculateSHA256(sectionValue);
+
+                if (embeddingGroup == null || !embeddingGroup.getEmbeddingSha256Hash().equals(sectionSha256Hash)) {
+
+
+                    embeddingResponse = embeddingService.getEmbeddingForMessage(projectAgent, sectionValue, projectAgent.getSemanticSearchModel());
+
+                } else {
+                    // If no training needed, use the existing embedding from the embedding group
+                    logger.debug("Embedding found in cache for section {}. Skipping...", item.getId());
+                    return null;
+                }
+            } else {
+                // If reuseEmbeddings is false, generate a new embedding
+                embeddingResponse = embeddingService.getEmbeddingForMessage(
+                        projectAgent, sectionValue, projectAgent.getSemanticSearchModel()
+                );
+            }
```

Comment:
This can be simplified to something like this:
```
if (Boolean.TRUE.equals(reuseEmbeddings)) {
    EmbeddingGroup embeddingGroup = embeddingService.findBySectionOrMessage(
        item.getId(), null, projectAgent.getSemanticSearchModel().getId()
    );

    String sectionSha256Hash = cryptographyUtils.calculateSHA256(sectionValue);

    if (embeddingGroup != null && embeddingGroup.getEmbeddingSha256Hash().equals(sectionSha256Hash)) {
        // Embedding found and hashes match; skip generating a new embedding
        logger.debug("Embedding found in cache for section {}. Skipping...", item.getId());
        return null;
    }
}

// Generate a new embedding if reuseEmbeddings is false or no matching embedding was found
embeddingResponse = embeddingService.getEmbeddingForMessage(
    projectAgent, sectionValue, projectAgent.getSemanticSearchModel()
);
```